tests/dialects/test_dlti.py

```diff
@@ -0,0 +1,181 @@
+    assert entry.key == StringAttr("test2")
+    assert entry.value == IntAttr(12)
+
+    # Test passing string for key and float for value
+    entry = DataLayoutEntryAttr("test3", FloatAttr(99.45, Float32Type()))
+    assert entry.key == StringAttr("test3")
+    assert entry.value == FloatAttr(99.45, Float32Type())
+
+    # Test passing type for key and string for value
+    entry = DataLayoutEntryAttr(i32, "test")
+    assert entry.key == i32
+    assert entry.value == StringAttr("test")
+
+    # Test passing StringAttr for key and value
+    entry = DataLayoutEntryAttr(StringAttr("k"), StringAttr("v"))
+    assert entry.key == StringAttr("k")
+    assert entry.value == StringAttr("v")
+
+
+def test_incorrect_data_layout_entry():
+    with pytest.raises(VerifyException):
+        DataLayoutEntryAttr("", "V")
+
+
+def generic_specification_test(
+    dlti_class: type[DLTIEntryMap],
+    comparison_entries: list[Attribute],
+):
+    # Comparison
+    comparison_attr = dlti_class.new(
+        (
+            ArrayAttr(
+                [
+                    DataLayoutEntryAttr("key_" + str(idx), e)
+                    for idx, e in enumerate(comparison_entries)
+                ]
+            ),
+        )
+    )
+
+    # Test initialisation from a dictionary
+    attr_one = dlti_class(
+        {"key_" + str(idx): e for idx, e in enumerate(comparison_entries)}
+    )
+    assert comparison_attr == attr_one
+    # Test initialisation from an array attribute of data layout entries
+    attr_two = dlti_class(
+        ArrayAttr(
+            [
+                DataLayoutEntryAttr("key_" + str(idx), e)
+                for idx, e in enumerate(comparison_entries)
+            ]
+        )
+    )
+    assert comparison_attr == attr_two
+
+
+@pytest.mark.parametrize(
+    "entries",
+    [
+        [StringAttr("value1"), IntAttr(23), IntAttr(43)],
+        [IntAttr(23), FloatAttr(2.4, Float32Type())],
+    ],
+)
+def test_data_layout_spec(
+    entries: list[Attribute],
+):
+    generic_specification_test(DataLayoutSpecAttr, entries)
+
+
+@pytest.mark.parametrize(
+    "entries",
+    [
+        [StringAttr("value1"), IntAttr(23), IntAttr(43)],
+        [IntAttr(23), FloatAttr(2.4, Float32Type())],
+    ],
+)
+def test_target_device_spec(
+    entries: list[Attribute],
+):
+    generic_specification_test(TargetDeviceSpecAttr, entries)
+
+
+@pytest.mark.parametrize(
+    "entries",
+    [
+        [StringAttr("value1"), IntAttr(23), IntAttr(43)],
+        [IntAttr(23), FloatAttr(2.4, Float32Type())],
+    ],
+)
+def test_target_system_spec(
+    entries: list[Attribute],
+):
+    generic_specification_test(TargetSystemSpecAttr, entries)
+
+
+@pytest.mark.parametrize(
+    "entries",
+    [
+        [StringAttr("value1"), IntAttr(23), IntAttr(43)],
+        [IntAttr(23), FloatAttr(2.4, Float32Type())],
+    ],
+)
+def test_map_attr(
+    entries: list[Attribute],
+):
+    generic_specification_test(MapAttr, entries)
```

Comment:
I would recommend using this two-step parametrization, which produces a product of the different parametrize entries:
```suggestion
@pytest.mark.parametrize(
    "cls", [DataLayoutSpecAttr, TargetDeviceSpecAttr, TargetSystemSpecAttr, MapAttr]
)
@pytest.mark.parametrize(
    "entries",
    [
        [StringAttr("value1"), IntAttr(23), IntAttr(43)],
        [IntAttr(23), FloatAttr(2.4, Float32Type())],
    ],
)
def test_target_device_spec(
    cls: type[DLTIEntryMap],
    entries: list[Attribute],
):
    generic_specification_test(TargetDeviceSpecAttr, entries)
```
Then I would recommend inlining the `generic_specification_test` function.

Reply:
Done :)